### PR TITLE
Fixed location property on TypedElementOnSourceNode.cs

### PR DESCRIPTION
**Pull request by Kasdejong**
## Description
TypedElementOnSourceNode.Location should now correctly omit choice type suffixes (like in PocoElementNode)

## Related issues
Fixes issue #2642 

## Testing
Added new unit tests testing some edge cases

### CODE REVIEW

File: src/Hl7.Fhir.Base/ElementModel/TypedElementOnSourceNode.cs [hunk 1]
```diff
@@ -31,6 +31,7 @@ public TypedElementOnSourceNode(ISourceNode source, string type, IStructureDefin
             if (source is IExceptionSource ies && ies.ExceptionHandler == null)
                 ies.ExceptionHandler = (o, a) => ExceptionHandler.NotifyOrThrow(o, a);
 
+            Location = source.Location;
```
Comment (ewoutkramer) [inline src/Hl7.Fhir.Base/ElementModel/TypedElementOnSourceNode.cs]: Why not source.Name if you copy the approach from shortpath?

File: src/Hl7.Fhir.ElementModel.Shared.Tests/TypedElementOnSourceNodeTests.cs [hunk 2]
```diff
@@ -22,5 +24,31 @@ public async Task TestExceptionComplexTypeValue()
 
             var _ = typedBundle.Children("entry").First().Value;
         }
+        
+        private SourceNode getAnnotatedNode => SourceNode.Valued("valueString", "hello");
+        
+        private SourceNode testPatient => SourceNode.Node("Patient",
+            SourceNode.Resource("contained", "Observation", SourceNode.Valued("valueBoolean", "true")),
+            SourceNode.Valued("active", "true",
+                getAnnotatedNode,
```
Comment (ewoutkramer) [inline src/Hl7.Fhir.ElementModel.Shared.Tests/TypedElementOnSourceNodeTests.cs]: This is unneccessary (and illegal)

**Review by ewoutkramer — CHANGES_REQUESTED** [review]

File: src/Hl7.Fhir.Base/ElementModel/TypedElementOnSourceNode.cs [hunk 3]
```diff
@@ -45,7 +48,7 @@ public TypedElementOnSourceNode(ISourceNode source, string type, IStructureDefin
                     throw Error.Format(nameof(type), $"Cannot determine the type of the root element at '{_source.Location}', " +
                         $"please supply a type argument.");
                 else
-                    return (rootType, null);
+                    return (rootType!, null);
```
Comment (ewoutkramer) [inline src/Hl7.Fhir.Base/ElementModel/TypedElementOnSourceNode.cs]: Let's consider this for a bit - for the first time we are kind of lying here, and we're now saying: "in 98% of the cases, this is non-null - so it's not null".  But is 98% good enough?  Or should we drop de bang and add a #pragma to make it clear we're dodging the nullability warning?

File: src/Hl7.Fhir.Base/ElementModel/TypedElementOnSourceNode.cs [hunk 4]
```diff
@@ -75,11 +78,12 @@ private TypedElementOnSourceNode(TypedElementOnSourceNode parent, ISourceNode so
             Definition = definition;
             InstanceType = instanceType;
             _settings = parent._settings;
+            Location = location;
         }
 
-        public ExceptionNotificationHandler ExceptionHandler { get; set; }
+        public ExceptionNotificationHandler ExceptionHandler { get; set; } = null!;
```
Comment (ewoutkramer) [inline src/Hl7.Fhir.Base/ElementModel/TypedElementOnSourceNode.cs]: Actually, this property *is* nullable I think?  It's possible to have no ExceptionHandler....

File: src/Hl7.Fhir.Base/ElementModel/TypedElementOnSourceNode.cs [hunk 5]
```diff
@@ -338,7 +342,7 @@ private string typeFromLogicalModelCanonical(ITypeSerializationInfo info)
             return pos > -1 ? type.Substring(pos + 1) : type;
         }
 
-        private bool tryGetBySuffixedName(Dictionary<string, IElementDefinitionSummary> dis, string name, out IElementDefinitionSummary info)
+        private bool tryGetBySuffixedName(Dictionary<string, IElementDefinitionSummary> dis, string name, out IElementDefinitionSummary? info)
```
Comment (ewoutkramer) [inline src/Hl7.Fhir.Base/ElementModel/TypedElementOnSourceNode.cs:345]: Shouldn't we have a NotNullWhen annotation here?

File: src/Hl7.Fhir.Base/ElementModel/TypedElementOnSourceNode.cs [hunk 6]
```diff
@@ -452,7 +459,7 @@ public IEnumerable<ITypedElement> Children(string name = null)
                 else
                     // Ok, pass through the untyped members, but since there is no type information, 
                     // don't bother to run the additional rules
-                    return enumerateElements(childElementDefs, _source, name);
+                    return enumerateElements(childElementDefs, _source, name!);
```
Comment (ewoutkramer) [inline src/Hl7.Fhir.Base/ElementModel/TypedElementOnSourceNode.cs]: `name` is nullable, and can actually be null, so why the bang?

**Review by ewoutkramer — CHANGES_REQUESTED** [review]

File: src/Hl7.Fhir.Base/ElementModel/TypedElementOnSourceNode.cs (hunk 5 above)
Comment (Kasdejong) [reply src/Hl7.Fhir.Base/ElementModel/TypedElementOnSourceNode.cs:345]: Screenshot 2024-03-26 at 20 37 09

Comment (Kasdejong) [reply src/Hl7.Fhir.Base/ElementModel/TypedElementOnSourceNode.cs:345]: I do not understand why it picks the newtonsoft here

Comment (Kasdejong) [reply src/Hl7.Fhir.Base/ElementModel/TypedElementOnSourceNode.cs:345]: ohhh, I know why! we have not merged the annotations branch yet :)

**Review by ewoutkramer — APPROVED** [review]
